### DISCUSSION TITLE
Using with start/end step ksamplers (such as in refiner workflows)?

### DISCUSSION THREAD

**Discussion by Jonseed** [Jonseed/ComfyUI-Detail-Daemon#4] [Q&A]
Some have asked if Detail Daemon or Lying Sigmas can be used in workflows with the advanced ksamplers, which have start and end steps values, which allow for refiners and such. I'm not sure it will work with those core nodes since they have the samplers built-in. It would require an advanced ksampler with the option to input a custom sampler. So far I haven't found a suitable custom node. Do you know of any @blepping ?

**Comment by blepping** [comment]
>  It would require an advanced ksampler with the option to input a custom sampler. So far I haven't found a suitable custom node. 

yeah, unfortunately you basically lose the ability to input parameters with the non-custom samplers.

i _probably_ don't know something that does what you're looking for, but also the question was pretty general so i'm not sure exactly what feature the user is looking for. there is one possible workaround (though it definitely has limitations). you can add samplers to the list users can choose from so you could do something like add a `euler_detail_daemon` (or whatever) with some reasonable preset values. that way people could choose it from the list, though they wouldn't be able to change parameters themselves.

i'd use that approach sparingly to avoid cluttering up the list with samplers only some people would actually want to use though, so probably keep it to 1-2 presets at most (unless you add configuration or something that lets the user specifically opt into it). also would be possible to do something like let the user create a config file in the custom node directory like some nodes do which allows defining some presets that can be added to the list when the node loads.

**Comment by Jonseed** [comment]
@blepping here is what they were looking for:

> "I wanted to know if its possible to use Detail daemon with the default Ksampler (Advanced), because I wanted to use "start at" and "end at" steps and "total steps" options. It seems that SamplerCustom doesn't have these options. I use two different ksamplers to control the step at which generation stops and then send noise to the other one, it's easier this way to control which lora to load and unload."

I haven't used that particular kind of workflow before with LoRAs, but I know SDXL uses a similar base/refiner workflow.

Impact Pack has a KSamplerAdvancedProvider node that looks like it lets you specify an input sampler, but I'm not sure this would work with the traditional KSamplerAdvanced with start and end steps...

**Reply by blepping** [reply]
>  because I wanted to use "start at" and "end at" steps and "total steps" options.

you can do that by slicing up the sigmas with a `SplitSigmas` node. 

i believe this would be the equivalent of start at step 2, end at step 4. 

![image](https://github.com/user-attachments/assets/1a32b526-62bb-4446-8833-59458013cc7e)

`high_sigmas` is the sigmas before the split, `low_sigmas` is after. 

*edit*: might be necessary to subtract one from the steps since `SplitSigmas` is zero based. not sure if the advanced sampler is or not. i.e. step 0 is the first step, step 1 is the second, etc.

**Reply by bluewanderer** [reply]
God! That's so brilliant!

**Comment by Jonseed** [comment]
I found a RegionalSamplerAdvanced that accepts a custom sampler, and has start and end steps. I wonder if this workflow could work.

![Screenshot 2024-10-31 103339](https://github.com/user-attachments/assets/0dbe9e66-5d11-4c7e-a479-602cfbda0a17)

**Comment by blepping** [comment]
actually, thinking about this some more (and running into my own situation where i wanted to use it but couldn't) i guess there is another alternative - make a model patch version. the user would need to pass it sigmas that match whatever the sampler is using _relatively_ closely (probably doesn't have to be exact because of the interpolation). i think they'd probably also have to manually set CFG.

ComfyUI's normal model patching stuff only works for SD1x/SD2x/SDXL though so at least the first pass would only support those models, no Flux or SD3x.

**Reply by Jonseed** [reply]
The only time I've thought about using this kind of thing is in SDXL base/refiner workflows where the generation passes from base to refiner samplers. But Detail Daemon is itself a kind of refiner, and upscaling with Flux does a better job of refining details too, so I have basically stopped using the SDXL refiner.

**Reply by blepping** [reply]
i can't tell from this response if you're interested in adding something like what i mentioned.

> But Detail Daemon is itself a kind of refiner, and upscaling with Flux does a better job of refining details too, so I have basically stopped using the SDXL refiner.

i'd say the main application would be more for stuff like FaceDetailer:

![image](https://github.com/user-attachments/assets/362652dc-2188-47c7-9fe7-c749507d45cc)

it can do stuff like detection internally, crop regions, inpaint faces (or whatever the detector can detect - hands, whatever) and then paste them back into the image. however, it does not support custom sampler input. it is a truly massive pain to do that sort of thing manually with a custom sampler.

**Reply by Jonseed** [reply]
So a model patch version would take the model and sigmas (from the sampler?) as input and output the adjusted model?

**Reply by blepping** [reply]
actually came up with a better option (well, revisited the preset option): https://github.com/blepping/ComfyUI-bleh#blehsetsamplerpreset

you can use the sampler presets associate something like DD or Lying Sigma Sampler to a item that can be selected from the list of samplers:

![image](https://github.com/user-attachments/assets/2183dde0-47a3-4350-bd25-5471525f939f)

this could be done with DD specific presets also but i think it makes more sense to add items to the list for a general feature that allows using custom samplers/setting their parameters even in nodes where it currently isn't possible.

**Reply by Jonseed** [reply]
This looks great! So, are any changes needed in DD or Lying Sigma, or can they be used with the BlehSetSamplerPreset as-is? What do you mean "add items to the list for a general feature"?

**Reply by blepping** [reply]
> So, are any changes needed in DD or Lying Sigma

shouldn't need any changes.

> What do you mean "add items to the list for a general feature"?

i meant it could be annoying for users to have a bunch of extra samplers they might not use added to the list (like DD-specific presets) but it makes more sense to do that for a general-purpose feature.

adding samplers to the list has to happen when the node pack gets loaded, not when the user actually uses a node in a workflow so whatever samplers get added there will be in the list regardless of whether the user actually wants/needs to use them.

did that answer your question?

**Reply by Jonseed** [reply]
Oh, ok. When you say "when the node pack gets loaded," do you mean when the ComfyUI frontend is loaded in the browser? So, if you set up a new preset, as you've shown, do you just reload the webpage to load it in the list?

To remove preset samplers from the list, do you just remove the BlehSetSamplerPreset from the workflow? Or how do you remove them?

Once you've set a preset, do you need to keep the BlehSetSamplerPreset node in your workflow, as you've shown, or can that be removed and the preset just selected in the sampler list?

**Reply by blepping** [reply]
>  So, if you set up a new preset, as you've shown, do you just reload the webpage to load it in the list?

nope, that node is just assigning something to the (pre-existing) preset 0.

> To remove preset samplers from the list, do you just remove the BlehSetSamplerPreset from the workflow? Or how do you remove them?

the number of presets in the list (whether or not they actually have a sampler associated to them) is determined when the ComfyUI backend loads the custom node pack - not when the user uses the node, refreshes their browser or whatever. you can select a sampler preset even if you've never added a `BlehSetSamplerPreset` node to your workflow - but you'll just get an error because that slot isn't populated.

so those slots in the sampler list will always be used regardless of whether the user is using/wants that feature. basically what i was talking about before - it wouldn't be a good precedent if a bunch of node packs added their own approach to presets to that list.

anyway, there's an environment variable `ComfyUI-bleh` looks at when it starts up to determine the number of presets to add to the sampler list (which defaults to 1 if unset). if you wanted to change the number of presets you'd need to change that environment variable in your Comfy startup scripts, restart the ComfyUI backend and refresh the page to pull in the new sampler lists.

> Once you've set a preset, do you need to keep the BlehSetSamplerPreset node in your workflow, as you've shown, or can that be removed and the preset just selected in the sampler list?

technically the node can be removed but i wouldn't recommend it - it's setting global state in the current running ComfyUI instance (the backend that is). so if you remove the node, the preset you associated will continue to work but if you restarted the backend then your workflow would break. setting a preset isn't adding anything to the workflow itself.

**Reply by Tastypotato0009** [reply]
Thank you! I followed your instructions, and now DD works with KSampler using BlehSetSamplerPreset!

**Reply by Jonseed** [reply]
Maybe I need to add some notes to the README about how to use DD and LyingSigmas with non-custom samplers, or at least point them to this discussion.

**Reply by Jonseed** [reply]
@blepping will the Multiply Sigmas node also work with the BlehSetSamplerPreset node, inputting the adjusted sigmas into the override_sigmas_opt?

**Reply by blepping** [reply]
> will the Multiply Sigmas node also work with the BlehSetSamplerPreset node, inputting the adjusted sigmas into the override_sigmas_opt?

it will work as in the feature will function as intended but it might not do what you want. those override sigmas _only_ apply to what gets passed to the sampler (not a node that does sampling like a `KSampler`, sampler as in Euler). here's an example of how you might run into unintended effects using multiplied sigmas (or different schedules/denoise levels) in the sigma override:

suppose you generate some sigmas and pass them to the preset sigma override: `1.1, 0.6, 0.0`

then you either pass sigmas to the `KSampler` (or whatever) or it generates them internally: `1.0, 0.5, 0.0`

The `KSampler` will add initial noise based on sigma `1.0`, but you're actually _sampling_ from sigma `1.1` since that's what got passed to the sampler so not enough noise will be added and you'll get less detail.

so the important thing to be aware of is if you override the sigmas like that, if the sampler node is starting from a different sigma there's probably going to be a mismatch between the amount of noise you're sampling with and the amount the sampler added. this is definitely not always bad: i often use a multiplier for the initial noise - just something to be aware of.

one way around it (which unfortunately doesn't work for RF models like Flux/SD3x) is to disable adding noise in the sampler and add it yourself using something like my `NoisyLatentLike` node from [ComfyUI-sonar](https://github.com/blepping/ComfyUI-sonar):

![image](https://github.com/user-attachments/assets/d4070342-cf39-49c9-9beb-79ed6b570a7a)

if you disable add noise, watch out for this bug which causes non-deterministic generations: https://github.com/comfyanonymous/ComfyUI/pull/4518 (bleh has a workaround for this which could be combined with the sampler preset - `BlehForceSeedSampler` will ensure the seed always gets set before sampling starts.)

**Reply by freshwind-bit** [reply]
Many thanks @Jonseed for the Detail Daemon Sampler and @blepping for such an elegant solution to integrate this node into a workflow with a variable step range of the chain of sequential advanced samplers (like in Dr. Lt. Data's video example with skipping one generation step and doubling the next one). Works great and is very effective both in terms of increasing and decreasing detail while maintaining the overall composition. Drastically game changer!

**Comment by 01Kingj** [comment]


**Reply by Jonseed** [reply]
Hmm, haven't seen that issue before. It might be your particular setup. Can you share the full error message?

**Comment by 01Kingj** [comment]
Thank you for taking time to respond, chat GPT to the rescue, it was a missing lib....  'ComfyUI\.venv\Scripts\python.exe -m pip install matplotlib'

Looks to be working okay now.

Sorry for taking up your time.



Jayson King



LinkedIn - LinkedIn Profile<https://www.linkedin.com/in/jayson-king-7245b33/>
________________________________
